src/event/event.service.ts

```diff
@@ -1,4 +1,84 @@
-import { Injectable } from '@nestjs/common';
+import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
+import { CreateEventDto } from './create-event.dto';
+import { PrismaService } from '../prisma/prisma.service';
+import { SupabaseService } from '../../supabase/supabase.service';
+import { isEmpty } from 'class-validator';
 
 @Injectable()
-export class EventService {}
+export class EventService {
+  constructor(
+    private readonly prismaService: PrismaService,
+    private readonly supabaseService: SupabaseService,
+  ) {}
+
+  async createEvent(data: CreateEventDto, files: Express.Multer.File[]) {
+    const {
+      name,
+      description,
+      registration_link,
+      start_time,
+      end_time,
+      location_id,
+    } = data;
+
+    try {
+      const conflictingEvent = await this.prismaService.event.findFirst({
+        where: {
+          location_id: location_id,
+          start_time: new Date(start_time),
+          end_time: new Date(end_time),
+        },
+      });
+
+      if (conflictingEvent) {
+        throw new HttpException(
+          'An event is already scheduled at this location at the same time',
+          HttpStatus.CONFLICT,
+        );
+      }
+
+      const event = await this.prismaService.$transaction(async (prisma) => {
+        const newEvent = await prisma.event.create({
+          data: {
+            name,
+            description,
+            registration_link,
+            start_time: new Date(start_time),
+            end_time: new Date(end_time),
+            location_id,
+          },
+        });
+
+        if (files && !isEmpty(files)) {
+          for (const file of files) {
+            const uniqueFileName = `${Date.now()}-${file.originalname}`;
+            const posterUrl = await this.supabaseService.uploadPosterToBucket(
+              file,
+            );
+
+            await prisma.poster.create({
+              data: {
+                event_id: newEvent.id,
+                image_url: posterUrl,
+              },
+            });
+          }
+        }
+
+        return newEvent;
+      });
```

Comment:
Can you revise this for query optimization lang. For example, we have 3 files to upload. Sa current implementation mo, file upload 2 will not start unless file upload 1 is finished, then file upload 3 will wait for file upload 2 to finish before starting the upload. To optimize this query we can batch the query para mas optimized sya. You can do something like this:

`if (files && !isEmpty(files)) {
  const uploadPromises = files.map(async (file) => {
    const uniqueFileName = `${Date.now()}-${file.originalname}`;
    const posterUrl = await this.supabaseService.uploadPosterToBucket(file);
    return { event_id: newEvent.id, image_url: posterUrl };
  });

  const posterData = await Promise.all(uploadPromises);
  await prisma.poster.createMany({ data: posterData });
}`